[Feat] 대댓글 수정 API 구현

jher235 opened the pull request:
## Related issue 🛠
- closed #152 
  


## Work Description 📝
- 대댓글 수정 API 구현
- 누락됐던 부분들 추가 - status 지정, final ...


## Uncompleted Tasks 😅

## To Reviewers 📢
- 갑자기 드는 생각인데, 현재 구조가 좀 객체지향적이지도 않고 데이터 지향적이지도 않은 것 같네요... 차라리 물리 FK를 걷어내거나 객체 수가 제한된 경우에는 양방향을 사용하는 방법도 좋겠다는 생각이 들었어요.

jher235 committed:
[Feat] #152 대댓글 status 추가

jher235 committed:
[Feat] #152 대댓글 이미지 status 추가

jher235 committed:
[Fix] #152 누락된 생성 시 sequence 부여하는 부분 추가

jher235 committed:
[Feat] #152 equals 및 hashcode 오버라이드

jher235 committed:
[Feat] #152 댓글 검증 로직 추가

jher235 committed:
[Feat] #152 sequence 필드 추가

jher235 committed:
[Feat] #152 대댓글 관련 예외 코드 작성

jher235 committed:
[Feat] #152 대댓글 조회 로직 구현

jher235 committed:
[Feat] #152 대댓글 update Dto 작성

jher235 committed:
[Feat] #152 대댓글 update 시 예외 상태 추가

jher235 committed:
[Feat] #152 대댓글 ACTIVE 이미지 목록 조회 로직 구현

jher235 committed:
[Feat] #152 대댓글 내용 update 메서드 작성

jher235 committed:
[Feat] #152 대댓글 이미지 로직 추가
- softDelete 및 시퀀스 업데이트 추가

jher235 committed:
[Feat] #152 대댓글 이미지 수정 비즈니스 로직 구현

jher235 committed:
[Feat] #152 대댓글 이미지 수정 presentation layer 구현

jher235 committed:
[Refactor] #152 관계 검증 로직 추출

jher235 committed:
[Fix] #152 누락된 final 추가

jher235 committed:
[Fix] #152 기본 status 값 할당

jeong1112 commented on src/main/java/org/sopt/bofit/domain/commentreply/repository/CommentReplyImageRepository.java:
```diff
@@ -1,10 +1,24 @@
 
+    List<CommentReplyImage> findAllByCommentReplyAndStatus(CommentReply commentReply, CommentReplyImageStatus status);
+
+    default Map<Long, CommentReplyImage> findAllByCommentReplyAndStatusAsMap(
+        CommentReply commentReply, CommentReplyImageStatus status
+    ){
+        return findAllByCommentReplyAndStatus(commentReply, status).stream()
+            .collect(Collectors.toMap(CommentReplyImage::getId, Function.identity()));
+    }
```
read 작업의 결과를 Map으로 변환하는 것은 repository의 책임이라고 보기는 어렵다고 생각하는데, 어떻게 생각하시나요?

jher235 replied:
되게 좋은 의견이네요~~ 근데 저는 이 부분이 비즈니스 로직인가??라는 의문이 있어요... 데이터를 <식별자, 객체>의 Map 형태로 조회하고 싶은 건데, 서비스 단에서는 repository의 내부 동작을 신경쓸 필요 없이 Map이라는 형태로 데이터를 조회하고 사용하도록 하기 위함이거든요. 마치 데이터를 Slice<> 형태로 반환해주는 것처럼요.

이건 관점에 따라서 다를 수 있을 것 같아요. 특히 저희 구조는 Service 하위의 Reader를 만들어두는 구조이기 때문에 이런 식의 데이터 가공 과정을 Reader에서 처리하는게 더 자연스럽게 느껴질 수도 있겠네요~

jher235 replied:
지금 이미지처리 Post 에서는 어떻게 하시나요?? S3에서 바로 삭제하세요??

jher235 replied:
좀 더 고민해봤는데, Reader 라는 이런 부분의 책임을 더욱 잘 가질 수 있는 계층이 존재하기 때문에 Reader에서 Map 으로 변환해서 데이터를 제공하는 구조로 가는 게 더 좋겠다는 생각이 들어서 수정했어요.
좋은 리뷰 감사합니다~~

jeong1112 replied:
지금은 바로 삭제하고 있어요 이미지를 굳이 복구할 여지가 있을까? 라는 생각이 들어서 그렇게 구현했습니다

jher235 replied:
지금 코드 상으로는 s3 삭제 로직이 아예 없고 그냥 DB 상의 데이터만 지우는 것 같은데요...?

소프트 딜리트를 제시하는 이유는 이미지 복구 가능성 + 신고 기능 등이 추가됐을 경우 이미지를 검수할 수 있도록 + 운영 환경에서 법률적 문제 등이 발생할 경우 대응 등의 이유입니다.
근데 지금 꼭 이런걸 지금 당장 만족할 필요는 없을 수 있어서 애매하긴 하네요...

jeong1112 replied:
아 맞아요 s3에서는 삭제를 안 하고 있어요 지금

jeong1112 replied:
그런 의미에서는 소프트 딜리트 방식도 괜찮아 보이는데, 소프트 딜리트 방식으로 게시물 이미지도 수정할게요

jeong1112 reviewed (approved):
고생하셨습니다~ 다만 지난 번에 얘기한 이미지 삭제 처리 방식에 대해서는 논의가 필요해 보여요

jher235 committed:
[Refactor] #152 Map 관련 변환을 Reader에서 진행하도록 변경
- 기존에는 Repository에서 데이터를 Map으로 반환받을 수 있도록. 데이터를 <식별자, 객체>의 Map 형태로 조회하고 서비스단에서는 이에 대해 신경쓰지 않도록. Repository에서 바로 반환할 수 있도록 default 메서드를 사용했었음. 데이터를 조회하는 부분에 대해서 비즈니스 로직으로 바라보고 싶지 않았던 것. 그러나 현재 서비스단의 구조상 Reader 라는 이런 부분의 책임을 더욱 잘 가질 수 있는 계층이 존재하므로 Reader 에서 Map으로 변환해서 데이터를 사용하도록 수정함

jher235 committed:
Merge branch 'develop' of https://github.com/team-bofit/bofit-server into feat/#152-update-comment-reply

# Conflicts:
#	src/main/java/org/sopt/bofit/domain/post/controller/PostController.java
